server/app/models/stack.rb

```diff
@@ -11,6 +11,8 @@ class Stack
   has_many :stack_revisions, dependent: :destroy
   has_many :stack_deploys, dependent: :destroy
   has_many :grid_services
+  has_many :volumes
+  embeds_many :external_volumes
```

Comment:
Agreed that the separate processing of `external_volumes` seems odd, but I see the point.. this is a complex N:M relationship, where the same `Volume` will either belong to a single `Stack`, or it might belong to a `Grid` (i.e. `volume.stack == nil`), but be referenced by many `Stack`s.

With some creative interpretation of the [Mongoid Relations : Referenced N-N](https://mongoid.github.io/old/en/mongoid/docs/relations.html#has_and_belongs_to_many) docs... might this work somehow?

```ruby
class Grid
  has_many :volumes
end

class Stack
  has_and_belongs_to_many :volumes
end

class Volume
  belongs_to :grid
  belongs_to :stack
end
```

This would then have something like a stack S with `{ "volume_ids": [ 'A', 'B' ]}` with grid-level volume A `{ "grid_id": "G", "stack_id": null}` and stack-level volume B `{ "grid_id": "G", "stack_id": "S" }`.... needs testing, I don't know if I understood this correctly :)